platforms/jvm/plugins-jvm-test-suite/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java

```diff
@@ -77,21 +74,24 @@ public DefaultJvmTestSuite(String name, SourceSetContainer sourceSets, Configura
         this.taskDependencyFactory = taskDependencyFactory;
         this.toolchainFactory = new ToolchainFactory(getObjectFactory(), getParentServices(), getInstantiatorFactory());
 
-        Configuration compileOnly = configurations.getByName(sourceSet.getCompileOnlyConfigurationName());
-        Configuration implementation = configurations.getByName(sourceSet.getImplementationConfigurationName());
-        Configuration runtimeOnly = configurations.getByName(sourceSet.getRuntimeOnlyConfigurationName());
-        Configuration annotationProcessor = configurations.getByName(sourceSet.getAnnotationProcessorConfigurationName());
-
         this.targets = getObjectFactory().polymorphicDomainObjectContainer(JvmTestSuiteTarget.class);
         this.targets.registerBinding(JvmTestSuiteTarget.class, DefaultJvmTestSuiteTarget.class);
 
-        this.dependencies = getObjectFactory().newInstance(
-                DefaultJvmComponentDependencies.class,
-                getObjectFactory().newInstance(DefaultDependencyAdder.class, implementation),
-                getObjectFactory().newInstance(DefaultDependencyAdder.class, compileOnly),
-                getObjectFactory().newInstance(DefaultDependencyAdder.class, runtimeOnly),
-                getObjectFactory().newInstance(DefaultDependencyAdder.class, annotationProcessor)
-        );
+        // This is a workaround for strange behavior from the Kotlin plugin.
+        // It seems Kotlin is not doing this anymore in the newest version, so we should re-evaluate
+        // whether we need withDependencies anymore.
+        //
+        // The Kotlin plugin attempts to look at the declared dependencies to know if it needs to add its own dependencies.
+        // We avoid triggering realization of getTestSuiteTestingFramework by only adding our dependencies just before
+        // resolution.
+        configurations.getByName(sourceSet.getImplementationConfigurationName())
+            .withDependencies(dep -> dep.addAllLater(getDependencies().getImplementation().getDependencies()));
+        configurations.getByName(sourceSet.getCompileOnlyConfigurationName())
+            .withDependencies(dep -> dep.addAllLater(getDependencies().getCompileOnly().getDependencies()));
+        configurations.getByName(sourceSet.getRuntimeOnlyConfigurationName())
+            .withDependencies(dep -> dep.addAllLater(getDependencies().getRuntimeOnly().getDependencies()));
+        configurations.getByName(sourceSet.getAnnotationProcessorConfigurationName())
+            .withDependencies(dep -> dep.addAllLater(getDependencies().getAnnotationProcessor().getDependencies()));
```

Comment:
Forgot to add...

Is this where we could use a `DependencySet.from(...)`?

Reply:
Yup this is where I was thinking. Except more like `DependencyScopeConfiguration.from` since `DependencySet` does not track constraints